src/engine/types/actions.rs

```diff
@@ -221,59 +221,82 @@ impl Display for MappingDeleteAction<Key> {
     }
 }
 
-/// A type for the return type of idempotent unlocking actions.
-pub struct SetUnlockAction<T> {
-    unlocked: Vec<T>,
+/// A type for the return type of idempotent unlocking actions on an
+/// individual pool.
+pub enum SetUnlockAction<T> {
+    /// If the pool was already started.
+    Identity,
+    /// If the unlock action resulted in the pool being started.
+    Started(Vec<T>),
```

Comment:
In general, I don't like designing data structures with two states that are essentially equivalent to `None` values. We actually had an issue up about designing the `Vec<_>`-based action structures this way, and I had a problem with it back then for this reason. I feel like having a single state that represents the `None` or empty value is always better for code readability reasons. I think this would be better if it followed the format of the other `Vec`-based action types.

Reply:
I recall the conversation. But here, this seemed best, because 1) the type returned by ```StratEngine::unlock_pool()``` must implement Display for the purposes of logging by the handle_action macro and 2) in the dbus_api ```unlock_pool()``` method that is being modified, it is necessary to distinguish between no devices unlocked because the pool is set up and no devices unlocked when the pool is not set up because it is necessary to know whether to modify the D-Bus properties or not.

There are ways to address this, like have ```StratEngine::unlock_pool``` return a tuple of say a bool and the unmodified SetUnlockAction struct and define Display for the tuple. But that doesn't seem like a better solution to me.

Another thing to do would be to skip the pre-check for whether the pool exists in ```StratEngine::unlock_pool()```.  In that case, though, even if a pool with the UUID exists, stratisd will end up trying to unlock and if succesfull set up the pool, even though it has the same UUID as an existing pool.

What about returning an Err in ```StratPool::unlock_pool()``` if another pool with the same UUID already exists? That is reasonably backward compatible and allows reverting the changes to  ```SetUnlockAction```. Because in previous versions of stratisd on UnlockPool, the udev events would be caught, stratisd would try to set up the pool, and then it would log a warning on failure in the background.

Reply:
That last suggestion isn't all that good, because it is only relevant in the highly unusual case where devices with the UUID exist in stopped pools _and_ there is a set up pool with the same UUID. Where there already is a set up pool _and_ there are no devices to unlock in the stopped pools, stratisd would previously return Ok with 0 unlocked UUIDs.

Reply:
I'm a little bit confused here. My understanding of the code is that an error is always returned if we attempt to start a pool and no devices are unlocked or even if some are unlocked and rolled back and the pool is not set up. I don't really understand why we need to differentiate between those two cases in the success (`Ok`) case mainly because I think the only time we'll get `Ok(_)` and no newly unlocked devices is if the pool is already set up. Can you explain the case you're thinking of in a little more detail where this distinction is necessary?

Reply:
@jbaublitz We could get ```<empty list>``` in the case where the pool is already started _and_ in the case where previously teardown failed to teardown any crypt devices (somehow) but on calling ```UnlockPool``` the pool is now started. In prior stratisd, in the second case, no attempt would be made to start the pool, because no udev events would occur (I think). In current situation, an attempt will be made to start the pool even if all devices are currently unlocked when ```PoolUnlock``` is called and the pool may be started.

We could modify behavior of start_pool() based on calling context such that it exits early, without actually starting the pool, with Ok if no devices are unlocked and it was invoked in unlock context.

In that case, no attempt would be made to start the pool and ```Ok(<empty list>)``` would always signify pool not started while ```Ok(<non-empty-list>)``` would always signify pool started and ```SetUnlockAction``` could keep its original definition.

Reply:
Okay! I think I understand what you're getting at now. Let me review this one more time. I may not ask for any changes and approve it now that I'm a little more clear on what you're getting at. I think you're right about this distinction being important and I can see why that case would not count as the identity portion of things if the pool is still newly started even if no new devices are unlocked.